doc/apcupsd.man
.\" manual page [] for apcupsd
.\" SH section heading
.\" SS subsection heading
.\" LP paragraph
.\" IP indented paragraph
.\" TP hanging label
.TH apcupsd 8 "APC UPS management under Linux -- November 1999"
.SH NAME
apcupsd \- daemon for most APC's UPS for Linux
.SH SYNOPSIS
.B /sbin/apcupsd
.br
.B /etc/apcupsd/apccontrol
.br
.B /etc/apcupsd/apcupsd.conf
.br
.B /sbin/apcaccess
.br
.B /sbin/apcnetd
.br
.SH DESCRIPTION
.LP
This document is considerably out of date. It can be used
for a brief overview of apcupsd, but you should consult the
html manual for current information. Normally this manual
will be loaded on your system in the doc directory, which
varies from system to system. Otherwise, you may find
an online version at: http://www.apcupsd.com

.LP
This daemon can be used for controlling the most APC UPSes.
During a power failure, "apcupsd" will inform the users about the
power failure and that a shutdown may occur.  If power is not
restored, a system shutdown will follow when the battery is
exausted, a timeout (seconds) expires, or runtime expires based
on internal UPS calculations determined by power consumption
rates.  If the power is restored before one of the above shutdown
conditions is met, apcupsd will inform users about this fact.

The shutdown is made by script calls to
"/etc/apcupsd/apccontrol", which is called by "apcupsd".
Consequently, no changes to /etc/inittab are necessary.  There is
no communication between apcupsd and init(1) process.  Apcupsd
modifies the halt script so that at the end of the shutdown
process, apcupsd will be re-executed in order to power off the
UPS. On certain platforms (e.g. FreeBSD) there is no halt script
so apccontrol must be modified to cause apcupsd to power off
the UPS after a delay time.

.LP
The apcupsd daemon now supports two networking modes that function
independently, but if desired they can both run at the same time.

Most users will probably enable the first network mode (NIS), which
permits apcupsd to serve status and event information 
to clients over the network.

The second networking mode is for multiple networked machines that are
powered by the same UPS. In this mode, one machine is configured as a
master with the UPS attached to the serial port.  The other machines
(max.  20) powered by the same UPS are configured as slaves.  The master
has a network connection with the slaves and sends them information
about the UPS status.

.LP
RedHat and SuSE versions of Linux have direct install support.  All
other flavors of Linux may need some fussing with to get the install
correct.
.SH OPTIONS
.TP
.B \-c --configure
Attempts to configure the UPS EPROM to the values specified in the
configuration file "/etc/apcupsd/apcupsd.conf".
This option is now deprecated (not supported) and should not
be used. Instead use the apctest program, which has an interactive
EEPROM configuration.
.TP
.B \-d --debug <level>
Turns on debugging output.
.TP
.B \-f --config-file <file>
Specifies the location of the configuration file.  The default
is: /etc/apcupsd/apcupsd.conf
.TP
.B \-k --killpower
Attempt to turn the UPS off. This option is normally only
used by the daemon itself to shut the UPS off after a
system shutdown has completed.
.TP
.B \-V --version
Prints the apcupsd version number and the help.
.TP
.B \-? --help
Prints a brief apcupsd help screen.

.SH CONFIGURATION
It may be necessary to change the configuration information in the file
"/etc/apcupsd/apcupsd.conf" to meet your needs and to correspond to your
configuration.  This file is a plain ASCII file and you can use your
favorite editor to change it. 
.LP 
Configuration commands in 
.B /etc/apcupsd/apcupsd.conf
are:

.LP
.SS GENERAL CONFIGURATION COMMANDS
.LP
.B  UPSCABLE
-- <type of cable you are using>
.br
[ simple | 940-0020B | 940-0023A (broken) ]
.br
[ smart | 940-0024B | 940-0024C ]
.br
[ 940-0095A | 940-0095C ]
.br
[ ether ]
.br 
[ usb ]
.LP
.B  UPSTYPE
-- <Type of APCC UPS you have>
.br
[ dumb | apcsmart | net | usb | snmp | test ] 
.LP
.B  DEVICE
-- <name of serial port>
.br
Please specify which device is used for UPS communication.
The default is /dev/ttyS[0|1|2|3].  If you have a USB UPS,
it is /dev/usb/hid/hiddev[0-9].  On some Red Hat versions
you may need to specify /dev/usb/hiddev[0-9]
.LP
.B  LOCKFILE
-- <path to lockfile>
.br
By supplying this argument, "apcupsd" tries to create a lockfile
for the serial or USB port in the specified directory.



.LP
.SS CONFIGURATION COMMANDS USED BY THE NETWORK INFORMATION SERVER
.LP
.B NETSTATUS 
[on | off]
.br
This configuration command turns the network information server
on or off. If it is on, apcupsd will spawn a child process that
serves STATUS and EVENTS information over the network. This
information is currently used by the Web based CGI programs.
The default is on. In some cases, for added security, you may
want to invoke a separate information server daemon from the inetd
daemon. In this case, NETSTATUS should be off.

.LP
.B STATUSPORT 
<port>
.br
This configuration command specifies the port to be used by the
apcupsd server. The default is 7000. If you change this port,
you must manually change the #define SERV_TCP_PORT in 
cgi/upsfetch.c and rebuild the CGI programs.

.LP
.B EVENTFILE 
<filename>
.br
If you want the apcupsd network information server to provide
the last 10 events via the network, you must specify a
file where apcupsd will save these events. The default is:
/var/log/apcupsd.events. Currently, apcupsd will save
at most the last 50 events. When more than 50 events are
recorded and a network request for the events arrives,
the network server will truncate the file to
the most recent 10 events. Consequently this file
will not grow indefinitely as long as the events are
checked from time to time.


.LP
.SS CONFIGURATION COMMANDS USED TO CONTROL SYSTEM LOGGING
.LP
.B STATTIME

.LP
.B STATFILE

.LP
.B DATATIME

.LP
.B FACILITY



.LP
.B CONFIGUATION COMMANDS USED DURING POWER FAILURES
.LP
.B  ANNOY
-- <time in seconds>
.br
Please specify the time in seconds between messages requesting
logged in users to get off the system. This timer starts only when
the UPS is running on batteries.  The default is 300 seconds 
(5 minutes).
.LP
.B  ANNOYDELAY
-- <time in seconds>
.br
Please specify delay time in seconds before apcupsd begins requesting
logged in users to get off the system. This timer starts
only after the UPS is running on batteries.
This timer is reset when the power returns. The default
is 60 seconds.  That is the first warning to log off the
system occurs after 60 seconds on batteries.


.LP
.B  NOLOGON
-- <specifies when apcupsd should create the nologon file>
.br
[ disable | timeout | percent | minutes | always ] are valid types.

Based on Ten (10) percent of a setting.  This allows one to define the
point when the /etc/nologin file is added.  This is important for
allowing systems with BIG UPSes to run as normally until the system
administrator determines the need for dumping users.  The feature also
allows the system administrator to hold the "ANNOY" factor until the
/etc/nologin file is added.

.B disable 
prevents apcupsd from creating the nologin file.

.B timeout 
specifies a specific wait time before creating the nologin file.

.B percent 
specifies the percent battery charge remaining before creating
the nologin file. 

.B minutes 
specifies the battery runtime remaining before creating the
nologin file.

.B always 
causes the nologin file to be immediately created on a power
failure.
.LP
.B  BATTERYLEVEL
-- <percent of battery>
.br
If BATTERYLEVEL is specified, during a power failure, apcupsd will
shutdown the system when the remaining battery charge falls below the
specified percentage.  The default is 5.
.LP
.B  MINUTES
-- <battery runtime in minutes>
.br
If MINUTES is specified, during a power failure, apcupsd will
shutdown the system when the remaining remaining runtime on batteries
as internally calculated by the UPS 
falls below the time specified. The default is 3.

.LP
.B  TIMEOUT
-- <time in seconds>
.br
After a power failure, the system will be shutdown after TIMEOUT seconds
have expired.  Normally for SMARTUPSes, this should be zero so that the
shutdown time will be determined by the battery level or remaining runtime 
(see above). This command is useful for dumb UPSes that do not report
battery level or the remaining runtime. It is also useful for
testing apcupsd in that you can force a rapid shutdown by setting
a small value (e.g. 60) and pulling the plug to the UPS. 

The timeout for the master is always 30 seconds longer than slaves.

.B TIMEOUT, BATTERYLEVEL, and MINUTES
can be set together without problems. The daemon will react to the
first case or test that is valid. Normally SmartUPS users will set
.B TIMEOUT
to zero so that the system is shutdown depending on the 
percentage battery charge remaining 
.B (BATTERYLEVEL)
or the remaining battery runtime
.B (MINUTES).


.LP
.SS CONFIGURATION COMMANDS FOR SHARING A UPS
.LP
.B  UPSCLASS
-- <class of operation>
.br
[ standalone | shareslave | sharemaster ] and
.br
[ netslave | netmaster ] are valid types.
.br
[ standalone | netslave | netmaster ] are tested classes.
.br
[ shareslave | sharemaster ] classes are being tested.
.LP
The default is "standalone" and should be used for 
all machines powered by the UPS and having a serial port
connection to the UPS, but where there are no other computers
dependent power from the same UPS.  This is the "normal" case.

Use "netmaster", if and only if you have a serial port connection
to the UPS and there are other machines deriving power from the
same UPS. 

Use "netslave" if and only if you have no serial port connection
to the UPS, but you derive power from it.

Use "shareslave" if and
only if you are using a ShareUPS and connected to a BASIC Port with
Simple Signal. 

Use "sharemaster", if and only if you are using a
ShareUPS and connected to the ADVANCED Port Smart Signal control.

.LP
.B  UPSMODE -- 
[ disable | share | net | sharenet ] are valid types.
.br
.LP
[ disable | net ] are the only known and tested classes.
.br
[ share | sharenet ] classes are being tested.
.LP
BETA [ share ] For two or seven (2/7) additional simple signal
ports on a SmartAccessories(tm) (internal/external box) for SmartUPSes.
.LP
.LP
.B  NETTIME
-- <time in seconds>
.br
The rate in seconds that broadcasts information to Slave machines.
This rate is reset if there is a power state change. This value
is now passed to the slaves to sync. the openning of a socket
by the slave for the master.
.LP
.B  NETPORT
-- <TCP|UDP port number>
.br
This unix service port number must be set in the /etc/services file
as follows:
.br
        tab     tab(spacejunk)  tab     tab
.br
name                    (stuff)/xxp             #
.br
.br
apcupsd         NETPORT/tcp             #
.br
apcupsd         NETPORT/udp             #
.LP
.B  MASTER
-- <name of master> for Slave machine.
.br
The name of the master which is authorized to send commands to this slave.
.LP
.B  SLAVE
-- <name of slave(s)> for Master machine.
.br
The name of the slave machine attached to the master.
There can be max. 20 slaves attached to one master.
.LP
.B  USERMAGIC
-- < user defined password> for the slave machine.
.br
The second level of password security. It must be (17) characters long
without spaces. This is passed to the master machine during
initialization of sockets. This string should be different for
each and every slave on the network. This is not at all secure as
passwords are passed in the clear. Please protect your network by
firewalling or tcpwrappers.

.SH SMARTUPS
If you start getting the follow message:
   
.B Emergency -- Batteries Have Failed!
.br
.B Change Them NOW!

Act upon it quickly. It means what it says.

Also, not all "SmartUPS" models are eqaully smart. A non-NET or old
class of "SmartUPS" has a subset of the full UPSlink(TM) language,
and can not be polled for its control codes.
.LP
.SH SHAREUPS
Special note for ShareUPS users,
.B TIMEOUT, BATTERYLEVEL, and MINUTES
are disabled or set to default values. Currently, there is not a known
way for early signals to be sent to BASIC Ports. MINUTES are set to 0.
.SH FILES
.B /etc/apcupsd/apcupsd.conf
- configuration file.
.br
.B /var/log/apcupsd.status
- STATUS file
.br
.B /var/log/apcupsd.events
- where up to the last 50 events are stored
for the network information server.
.SH EVENTS
apcupsd generates events when certain conditions occur 
such as a power failure, batteries exhausted, power return, 
...

These events are sent to the system log, optionally sent
to the temporary events file (/var/log/apcupsd.events),
and they also generate a call to /etc/apcupsd/apccontrol
which in turn will call any scripts you have placed in the
/etc/apcupsd directory.

.SH DATA format
If the DATATIME configuration command is set nonzero, 
apcupsd will log a data record at the interval defined
on the DATATIME command. This data record is in a format
similar to the PowerChute data file format.

.SH STATUS format
The STATUS output is in ASCII format and generally there
is a single piece of information on each line output.
The format varies based on the type of UPS that you are using.
.LP
.br
DATE     : time and date of last update
.br
CABLE    : cable type used
.br
UPSMODEL : ups type or signal method
.br
UPSMODE  : tells apcupsd what to check
.br
SHARE    : if ShareUPS is used, this determines what
.LP
.B SmartUPS and MatrixUPS Smart Signals
.br
ULINE    : Current (observed) Input Line Voltage
.br
MLINE    : Max (observed) Input Line Voltage
.br
NLINE    : Min (observed) Input Line Voltage
.br
FLINE    : Line Freq (cycles)
.br
VOUTP    : UPS Output Voltage
.br
LOUTP    : Percent Load of UPS Capacity
.br
BOUTP    : Current Charge Voltage of Batteries
.br
BCHAR    : Batteries Current Charge Percent of Capacity
.br
BFAIL    : UNSIGNED INT CODE (ups state)
.br
UTEMP    : Current UPS Temp. in Degrees Cel.
.br
DIPSW    : Current DIP switch settings for UPS.
.LP
.B Newer BackUPS Pro Smart Signals
.br
ULINE    : Current (observed) Input Line Voltage
.br
MLINE    : Max (observed) Input Line Voltage
.br
NLINE    : Min (observed) Input Line Voltage
.br
FLINE    : Line Freq (cycles)
.br
VOUTP    : UPS Output Voltage
.br
LOUTP    : Percent Load of UPS Capacity
.br
BOUTP    : Current Charge Voltage of Batteries
.br
BCHAR    : Batteries Current Charge Percent of Capacity
.br
BFAIL    : UNSIGNED INT CODE (ups state)
.LP
.B BackUPS Pro and SmartUPS v/s Smart Signals
.br
LINEFAIL : OnlineStatus
.br
BATTSTAT : BatteryStatus
.br
MAINS    : LineVoltageState
.br
LASTEVNT : LastEventObserved
.LP
.B BackUPS and NetUPS Simple Signals
.br
LINEFAIL : OnlineStatus
.br
BATTSTAT : BatteryStatus
.LP
.B BackUPS Pro and SmartUPS v/s Smart Signals
.br
OnlineStatus BatteryStatus LineVoltageState
LastEventObserved
.LP
.B BackUPS and NetUPS Simple Signals
.br
OnlineStatus BatteryStatus
.br

.SH SEE ALSO
The html manual installed on your system or
http://www.apcupsd.com

.SH AUTHOR
Andre M. Hedrick
.SS Retired Co-AUTHOR
Christopher J. Reimer
.SS Former programmer
Riccardo Fachetti
.SS Current maintainer
Kern Sibbald
.SS Contributors
An enormous list of past and former persons who have devoted
their time and energy to this project -- thanks.


ChangeLog
/***************************************************************************/
/*                     ChangeLog of Apcupsd 3.12.x                         */
/*                                                                         */
/*               Adam Kropelin <akropel1 at rochester dot rr dot com>      */
/*               Kern Sibbald <kern at sibbald dot com>                    */
/*                                                                         */
/*               http://www.apcupsd.com/                                   */
/*               http://sourceforge.net/projects/apcupsd                   */
/***************************************************************************/

CHANGES IN 3.12.x

13Feb06:
- Don't link against $(DRVLIBS) unless it is necessary. This fixes a link
  error in apcaccess on AIX 5.1.
- apcsmart: Add debug output to track data returned by UPS.
11Feb06:
- bsd-usb: Increase size of interrupt report buffer. This works around a 
  problem where the UPS would deliver several interrupt reports at once, 
  and they would not evenly fill the buffer. Since we only look at the 
  first report in the buffer, the next time around we'd see the tail of a 
  previous report and parse it as a new report, leading to momentary 
  glitches of mangled data.
09Feb06:
- bsd-usb: Fix bug masking 32 bit fields. Oddly, the linux kernel had this
  exact same bug a few months ago.
- Introduce logf() and use it to ensure free-form debug prints make it to
  the trace file when using -T.
02Feb06:
- net: Add support for selftest detection.

CHANGES IN 3.12.2

17Jan06:
- win32 Use OpenThreadToken() instead of OpenProcessToken(). This makes
  shutdown work on WinXP.
08Jan06:
- win32: Fix typo so Win32 event window will be populated correctly.
- win32: Include sh.exe and requisite DLLs.
- win32: Fix typo in battattach and battdetach cases.


CHANGES IN 3.12.1

06Jan06:
- usb: Status bits must use positive logic, so online state must be 
  updated even if CI_Discharging=0. This fixes missing status on SU2200.


CHANGES IN 3.12.0

02Jan06:
- doc: Update version to 3.12.x and remove ChangeLog and ReleaseNotes
  sections since it is better to fetch them directly from SourceForge
  or from the source tarball.
01Jan06:
- snmp: Remove misuse of NETSNMP_CALLBACK_OP_RECEIVED_MESSAGE. Also use
  snmp_{session,pdu} instead of netsmp_{session,pdu} to retain 
  compatibility with the ucd-snmp library.
- Fix compliation of examples/ files: client.c, megaclient.c, 
  newslave.c, upsapm.c  
31Dec05:
- conf: Update EEPROM section comments.
19Dec05:
- win32: Convert Readme.TXT to DOS text format so it will be readble 
  in Notepad (default *.txt handler on Windows). Remove version number 
  since I'll never remember to update it. Point users to manual for 
  detailed install instructions.
- m/s: Fix sense of onbatt/online, which was reversed.
16Dec05:
- doc: Fix latex errors that crept in during SNMP and USB updates.
- configure: Force NIS to enabled since libnis is required by 
  apcaccess. Users may still disable NIS at runtime in apcupsd.conf.
- doc: Remove references to --enable-nis and --disable-nis.
- doc: Update Win32 section to cover nifty new NSIS installer.
15Dec05:
- win32: More installer fixes: Create start/stop links regardless of 
  service or not. Remove apcupsd.events on uninstall, if user wants to.
  Fix compile error in installation directory message text.
- win32: A few last fixes (hopefully): WinNT and 9x need different
  service start methods. Remove extra double-quotes around Win9x start 
  handling because they seem to result in a bogus Exec commandline.
14Dec05:
- win32: Fix an uninstaller bug where apcupsd.conf was left behind 
  even if the user chose to remove it.
- win32: Fix uninstall so it doesn't miss any files. Change destination
  path instructions to help prevent inoperable installations. Fix 
  documentation and start link creation.
26Nov05
- Update the version
- Cleanup errors in the uninstall script
- Remove a few MessageBoxes that annoy on Win98
- Set default UPSTYPE to apcsmart (was smartups)
- Add examples to installer script.
- Try to remove all files on uninstall.
- Add "missing" installer macro files.
- Implement NSIS installer. It can be downloaded from:
  http://sourceforge.net/projects/nsis/  I'm using version
  2.0, but the latest one is 2.11
- Create winapcupsd.nsi.in file in src/win32 directory.
- Modify configure.in to generate winapcupsd.nsi file.
- Remove most of the old ServiceHelper code. It is not used,
  and creates problems on remote logins.
- Add nagios plugin contributed by Christian Masopust.
- Add all the new cygwin 1.5.18 binaries and the dll.
- Eliminate some compiler warnings.
- Make the old make binary-release work.
- Implement a new target for building the NSIS installer.
    ./configure (options...)
    make
    cd src/win32
    make winapcupsd.exe
24Nov05:
- Get rid of unsupported -c (configure), -u (update battery date), 
  and -n (rename) options.
22Nov05:
- configure: -mwindows is only needed for the link step, so kill
  $WCFLAGS and let all compile steps be driven by the .c.o rule.
- configure: Don't add ncurses libs to $LIBS, only to $POWERLIBS.
  That prevents libncurses from being linked with every executable.
- configure: Kill apcnisd. It requires sysv shmem and since we've 
  eliminated that, so too must apcnisd be eliminated.
19Nov05:
- snmp: Initialize nfds to zero before calling snmp_select_info().
  Fixes crash on Win32 and potential crash on other platforms.
- win32: Build fixes: apc_defines.h -> defines.h, remove use of 
  astr* pending future cleanup, C++ casting fixes.
- win32: Switch to pthreads for the Main_Msg_Loop thread. This
  seems to fix a bug where Main_Msg_Loop called do_status() which
  attempted to take the UPS read_lock. Apparently taking pthreads 
  locks from CreateThread threads is bad. Switching to pthreads
  seems to fix it.
12Nov05:

tags/Release-/src/apcupsd.c
/*
 * apcupsd.c
 *
 * APC UPS monitoring daemon.
 */

/*
 * Copyright (C) 1999-2005 Kern Sibbald
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General
 * Public License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
 * MA 02111-1307, USA.
 */

/* The big overall flow of apcupsd is as follows
 *
 * Assuming no networking features are turned on.
 *
 * Main Process:
 *   process configuration file
 *   establish contact with the UPS
 *   read the state of the UPS
 *   Fork to become a daemon
 *   write the shared memory
 *   start child process apcdev -- subroutine do_device()
 *   start child process apcnis, if configured -- subroutine do_server()
 *   wait for child processes to complete
 *   exit
 *
 * Child process apcdev -- subroutine do_device()
 *   wait for activity on serial port or timeout 
 *        from select() (5 seconds) -- check_serial()
 *   call do_action()
 *        read and lock the shared memory
 *        check to see if UPS state has changed and if we should take any action. I.e.
 *              are we on batteries, should we shutdown, ...
 *        write and unlock shared memory
 *   call fillUPS()
 *        read and lock shared memory
 *        Poll the UPS for each state variable
 *        enable the UPS
 *        enable the UPS again
 *        write and unlock shared memory
 *   call do_reports()
 *        if DATA report time expired, write DATA report -- log_data()
 *        if STATUS report time expired, write STATUS report
 *   loop
 */

#include "apc.h"

/*
 * myUPS is a structure that need to be defined in _all_ the forked processes.
 * The syncronization of data into this structure is done with the shared
 * memory area so this is made reentrant by the shm mechanics.
 */
UPSINFO *core_ups = NULL;

static void daemon_start(void);

int pidcreated = 0;
extern int kill_on_powerfail;
extern FILE *trace_fd;
extern char *pidfile;

/*
 * The terminate function and trapping signals allows apcupsd
 * to exit and cleanly close logfiles, and reset the tty back
 * to its original settings. You may want to add this
 * to all configurations.  Of course, the file descriptors
 * must be global for this to work, I also set them to
 * NULL initially to allow terminate to determine whether
 * it should close them.
 */
void apcupsd_terminate(int sig)
{
   UPSINFO *ups = core_ups;

   if (sig != 0)
      log_event(ups, LOG_WARNING, "apcupsd exiting, signal %u", sig);

   clean_threads();
   clear_files();
   if (ups->driver)
      device_close(ups);
   delete_lockfile(ups);
   if (pidcreated)
      unlink(pidfile);
   log_event(ups, LOG_WARNING, "apcupsd shutdown succeeded");
   destroy_ups(ups);
   closelog();
   _exit(0);
}

void apcupsd_error_cleanup(UPSINFO *ups)
{
   if (ups->driver)
      device_close(ups);
   delete_lockfile(ups);
   if (pidcreated)
      unlink(pidfile);
   clean_threads();
   log_event(ups, LOG_ERR, "apcupsd error shutdown completed");
   destroy_ups(ups);
   closelog();
   exit(1);
}

/*
 * Subroutine error_out prints FATAL ERROR with file,
 * line number, and the error message then cleans up 
 * and exits. It is normally called from the Error_abort
 * define, which inserts the file and line number.
 */
void apcupsd_error_out(const char *file, int line, const char *fmt, ...)
{
   char buf[256];
   va_list arg_ptr;
   int i;

   asnprintf(buf, sizeof(buf),
      "apcupsd FATAL ERROR in %s at line %d\n", file, line);

   i = strlen(buf);
   va_start(arg_ptr, fmt);
   avsnprintf((char *)&buf[i], sizeof(buf) - i, (char *)fmt, arg_ptr);
   va_end(arg_ptr);

   fprintf(stderr, "%s", buf);
   log_event(core_ups, LOG_ERR, "%s", buf);
   apcupsd_error_cleanup(core_ups);     /* finish the work */
}

/*
 * Subroutine error_exit simply prints the supplied error
 * message, cleans up, and exits.
 */
void apcupsd_error_exit(const char *fmt, ...)
{
   char buf[256];
   va_list arg_ptr;

   va_start(arg_ptr, fmt);
   avsnprintf(buf, sizeof(buf), (char *)fmt, arg_ptr);
   va_end(arg_ptr);

   fprintf(stderr, "%s", buf);
   log_event(core_ups, LOG_ERR, "%s", buf);
   apcupsd_error_cleanup(core_ups);     /* finish the work */
}

/*
 * ApcupsdMain is called from win32/winmain.cpp
 * we need to eliminate "main" as an entry point,
 * otherwise, it interferes with the Windows
 * startup.
 */
#ifdef HAVE_MINGW
# define main ApcupsdMain
#endif

/* Main program */
int main(int argc, char *argv[])
{
   UPSINFO *ups;
   int tmp_fd, i;

   /* Set specific error_* handlers. */
   error_out = apcupsd_error_out;
   error_exit = apcupsd_error_exit;

   /*
    * Default config file. If we set a config file in startup switches, it
    * will be re-filled by parse_options()
    */
   cfgfile = APCCONF;

   /*
    * Create fake stdout in order to circumvent problems
    * which occur if apcupsd doesn't have a valid stdout
    * Fix by Alexander Schremmer <alex at alexanderweb dot de>
    */
   tmp_fd = open("/dev/null", O_RDONLY);
   if (tmp_fd > 2) {
      close(tmp_fd);
   } else {
      for (i = 1; tmp_fd + i <= 2; i++)
         dup2(tmp_fd, tmp_fd + i);
   }

   /*
    * If there's not one in libc, then we have to use our own version
    * which requires initialization.
    */
   strlcpy(argvalue, argv[0], sizeof(argvalue));

   ups = new_ups();                /* get new ups */
   if (!ups)
      Error_abort1("%s: init_ipc failed.\n", argv[0]);

   init_ups_struct(ups);
   core_ups = ups;                 /* this is our core ups structure */

   /* parse_options is self messaging on errors, so we need only to exit() */
   if (parse_options(argc, argv))
      exit(1);

   Dmsg(10, "Options parsed.\n");

   if (show_version) {
      printf("apcupsd " APCUPSD_RELEASE " (" ADATE ") " APCUPSD_HOST "\n");
      exit(0);
   }

   /*
    * We open the log file early to be able to write to it
    * it at any time. However, please note that if the user
    * specifies a different facility in the config file
    * the log will be closed and reopened (see match_facility())
    * in apcconfig.c.  Any changes here should also be made
    * to the code in match_facility().
    */
   openlog("apcupsd", LOG_CONS | LOG_PID, ups->sysfac);

#ifndef DEBUG
   if ((getuid() != 0) && (geteuid() != 0))
      Error_abort0("Needs super user privileges to run.\n");
#endif

   check_for_config(ups, cfgfile);
   Dmsg(10, "Config file %s processed.\n", cfgfile);

   /*
    * Disallow --kill-on-powerfail in conjunction with simple signaling
    * UPSes. Such UPSes have no shutdown grace period so using --kill-on-
    * powerfail would guarantee an unclean shutdown.
    */
   if (kill_on_powerfail && ups->mode.type == DUMB_UPS) {
      kill_on_powerfail = 0;
      log_event(ups, LOG_WARNING,
         "Ignoring --kill-on-powerfail since it is unsafe "
            "on Simple Signaling UPSes");
   }

   /* Attach the correct driver. */
   attach_driver(ups);
   if (ups->driver == NULL)
      Error_abort0("Apcupsd cannot continue without a valid driver.\n");

   ups->start_time = time(NULL);

   if (!hibernate_ups && !shutdown_ups && go_background) {
      daemon_start();

      /* Reopen log file, closed during becoming daemon */
      openlog("apcupsd", LOG_CONS | LOG_PID, ups->sysfac);
   }

   init_signals(apcupsd_terminate);

   /* Create temp events file if we are not doing a hibernate or shutdown */
   if (!hibernate_ups && !shutdown_ups && ups->eventfile[0] != 0) {
      ups->event_fd = open(ups->eventfile, O_RDWR | O_CREAT | O_APPEND, 0644);
      if (ups->event_fd < 0) {
         log_event(ups, LOG_WARNING, "Could not open events file %s: %s\n",
            ups->eventfile, strerror(errno));
      }
   }

   if (create_lockfile(ups) == LCKERROR) {
      Error_abort0("Unable to create UPS lock file.\n"
                   "  If apcupsd or apctest is already running,\n"
                   "  please stop it and run this program again.\n");
   }

   make_pid_file();
   pidcreated = 1;

   if (hibernate_ups || shutdown_ups)
   {
      // If we're hibernating or shutting down the UPS, setup is a one-shot.
      // If it fails, we're toast; no retries
      if (!setup_device(ups))
         Error_abort0("Unable to open UPS device for hibernate or shutdown");

      if (hibernate_ups)
         initiate_hibernate(ups);
      else
         initiate_shutdown(ups);

      apcupsd_terminate(0);
   }

   /*
    * From now ... we must _only_ start up threads!
    * No more unchecked writes to myUPS because the threads must rely
    * on write locks and up to date data in the shared structure.
    */

   /* Network status information server */
   if (ups->netstats) {
      start_thread(ups, do_server, "apcnis", argv[0]);
      Dmsg(10, "NIS thread started.\n");
   }

   log_event(ups, LOG_WARNING,
      "apcupsd " APCUPSD_RELEASE " (" ADATE ") " APCUPSD_HOST " startup succeeded");

   /* main processing loop */
   do_device(ups);

   apcupsd_terminate(0);
   return 0;                       /* to keep compiler happy */
}

extern int debug_level;

/*
 * Initialize a daemon process completely detaching us from
 * any terminal processes.
 */
static void daemon_start(void)
{
#if !defined(HAVE_WIN32)
   int i, fd;
   pid_t cpid;
   mode_t oldmask;

#ifndef HAVE_QNX_OS
   if ((cpid = fork()) < 0)
      Error_abort0("Cannot fork to become daemon\n");
   else if (cpid > 0)
      exit(0);                     /* parent exits */

   /* Child continues */
   setsid();                       /* become session leader */
#else
   if (procmgr_daemon(EXIT_SUCCESS, PROCMGR_DAEMON_NOCHDIR
                                    | PROCMGR_DAEMON_NOCLOSE
                                    | PROCMGR_DAEMON_NODEVNULL
                                    | PROCMGR_DAEMON_KEEPUMASK) == -1)
   {
      Error_abort0("Couldn't become daemon\n");
   }
#endif   /* HAVE_QNX_OS */

   /* Call closelog() to close syslog file descriptor */
   closelog();

   /*
    * In the PRODUCTION system, we close ALL file descriptors unless
    * debugging is on then we leave stdin, stdout, and stderr open.
    * We also take care to leave the trace fd open if tracing is on.
    * Furthermore, if tracing we redirect stdout and stderr to the
    * trace log.
    */
   for (i=0; i<sysconf(_SC_OPEN_MAX); i++) {
      if (debug_level && i == STDIN_FILENO)
         continue;
      if (trace_fd && i == fileno(trace_fd))
         continue;
      if (debug_level && (i == STDOUT_FILENO || i == STDERR_FILENO)) {
         if (trace_fd)
            dup2(fileno(trace_fd), i);
         continue;
      }
      close(i);
   }

   /* move to root directory */
   chdir("/");

   /* 
    * Avoid creating files 0666 but don't override a
    * more restrictive umask set by the caller.
    */
   oldmask = umask(022);
   oldmask |= 022;
   umask(oldmask);

   /*
    * Make sure we have fd's 0, 1, 2 open
    * If we don't do this one of our sockets may open
    * there and if we then use stdout, it could
    * send total garbage to our socket.
    */
   fd = open("/dev/null", O_RDONLY);
   if (fd > 2) {
      close(fd);
   } else {
      for (i = 1; fd + i <= 2; i++)
         dup2(fd, fd + i);
   }
#endif   /* HAVE_WIN32 */
}
